File: api-call-limit/atomic/src/main/java/oscar/answers/service/CallCounter.java
package oscar.answers.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

@Component
public class CallCounter {

    private Logger logger;
    private AtomicInteger counts;
    private Long expireTime;

    @Autowired
    public CallCounter() {
        this.logger = LoggerFactory.getLogger(CallCounter.class);
        this.counts = new AtomicInteger();
        this.expireTime = System.currentTimeMillis();
    }

    @Scheduled(fixedRate = 1000)
    public void removeCountAfterOneMinute() {
        if (expireTime + 60000L <= System.currentTimeMillis()) {
            expireTime = System.currentTimeMillis();
            counts.set(0);
        }
    }

    public List<?> getCountAndExpiretime() {
        return List.of(counts, expireTime);
    }

    public int get() {
        return counts.incrementAndGet();
    }
}

File: kotlin/api-call-limit/atomic/src/main/kotlin/oscar/kotlin/answers/service/CallCounter.kt
package oscar.kotlin.answers.service

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.util.concurrent.atomic.AtomicInteger

@Component
class CallCounter {

    private val logger: Logger = LoggerFactory.getLogger(CallCounter::class.java)
    private val counts: AtomicInteger = AtomicInteger()
    private var expireTime: Long = System.currentTimeMillis()

    @Scheduled(fixedRate = 1000)
    fun removeCountAfterOneMinute() {
        if (expireTime + 60000L <= System.currentTimeMillis()) {
            expireTime = System.currentTimeMillis()
            counts.set(0)
        }
    }

    fun getCountAndExpiretime(): List<*> {
        return listOf(counts, expireTime)
    }

    fun get(): Int {
        return counts.incrementAndGet()
    }
}